src/Chromatogram.js

```diff
@@ -218,6 +220,111 @@ export class Chromatogram {
     return applyLockMass(this, mfs, options);
   }
 
+  getMatrix(range = []) {
+    let matrices = [];
+    for (let p = 0; p < range.length; p++) {
+      let peak = range[p];
+      let { from, to } = peak;
+      let fromIndex = this.getClosestTime(from);
+      let toIndex = this.getClosestTime(to);
+
+      let data = this.series.ms.data.slice(fromIndex, toIndex);
+      let times = this.times.slice(fromIndex, toIndex);
+
+      let xAxis = new Set();
+      for (let i = 0; i < data.length; i++) {
+        let spectrum = data[i];
+        for (let j = 0; j < spectrum[0].length; j++) {
+          xAxis.add(Math.round(spectrum[0][j]));
+        }
+      }
+      xAxis = Array.from(xAxis).sort((a, b) => a - b);
+      const nbPoints = xAxis.length;
+      const matrix = new Array(data.length);
+      for (let i = 0; i < data.length; i++) {
+        let element = data[i];
+        let row = new Float32Array(nbPoints);
+        for (let j = 0; j < element[0].length; j++) {
+          let xValue = Math.round(element[0][j]);
+          let yValue = element[1][j];
+          let index = xFindClosestIndex(xAxis, xValue);
+          row[index] = yValue;
+        }
+        matrix[i] = Array.from(row);
+      }
+      matrices.push({ times, xAxis, matrix });
+    }
+    return matrices;
+  }
+
+  estimateNbPureComponents(range = []) {
+    let ranks = [];
+    for (let p = 0; p < range.length; p++) {
+      let matrix;
+      let data = range[p];
+      if (data.matrix) {
+        matrix = data.matrix;
+      } else {
+        let result = this.getMatrix([data]);
+        matrix = result[0].matrix;
+      }
+      let pca = new PCA(matrix, {
+        method: 'NIPALS',
+        nCompNIPALS: 10,
+        scale: true,
+        ignoreZeroVariance: true,
+      });
+      let s = pca.getExplainedVariance();
+      let i = 1;
```

Comment:
Why having this outside de for loop?

Reply:
If do you mean the i variable, it is outside because of I need to use it outside. I'm open to any suggestion

Reply:
For me a while loop seems more natural, but is equally valid
```js
let index = 1;
let cumulative = s[0];
while((cumulative - s[index]) / cumulative > 0.88 && index < s.length) {
  cumulative += s[i];
  index++;
}
ranks.push(index);
```